File: ranvar/digest.py
"""
Contains Digest classes that approximates quantiles of data streams using
the t-digest algorithm. The t-digest algorithm is a probabilistic data structure
that can be used to estimate the quantiles of a data stream with a small memory footprint.
"""


from types import prepare_class

import numpy as np
import cython as pyx

if pyx.compiled:
    from cython.cimports.libc.math import round as c_round
    from cython.cimports.libc.math import ceil as ceil
    from cython.cimports.libc.math import floor as floor
    from cython.cimports.libc.time import time as c_time
    from cython.cimports.libc.stdlib import rand as c_rand
    from cython.cimports.libc.stdlib import srand as c_srand
    from cython.cimports.libc.stdlib import RAND_MAX as c_RAND_MAX
    # from cython.cimports.random import normal
else:
    ceil = np.ceil
    floor = np.floor



if pyx.compiled:
    print('Running through Cython!')
else:
    print('WARNING: Not Compiled.')


# c_srand(c_time(pyx.NULL))

#---[ Probability Distributions ]-----------------------------------------------

# # Rand
# @pyx.cfunc
# def _rand() -> pyx.double:
#     out:pyx.double = pyx.cast(pyx.double, c_rand()) / pyx.cast(pyx.double, c_RAND_MAX)
#     return out

# # RandInt
# @pyx.cfunc
# def _randint(l: pyx.double, h: pyx.double) -> pyx.double:
#     return c_round((h - l) * _rand() + l)



# TDigest


LOWER = (1 - 0.9999) / 2
UPPER = 1 - LOWER

__ADD__:pyx.int = 0
__MUL__:pyx.int = 1
__MAX__:pyx.int = 2
__MIN__:pyx.int = 3
__SUB__:pyx.int = 4
__POW__:pyx.int = 5

DEFAULTS = {
    'maxBins':32,
    'compress':True,
    'tolerance':1e-3
}




@pyx.cclass
class TDigest():
    _bins: pyx.double[:]
    _cnts: pyx.double[:]

    bins: np.ndarray
    cnts: np.ndarray

    maxBins: pyx.int
    nActive: pyx.int

    def __init__(self, maxBins=32):
        self.maxBins = maxBins
        self.nActive = 0

        self.bins = np.zeros(self.maxBins + 1, dtype=np.float64)
        self.cnts = np.zeros(self.maxBins + 1, dtype=np.float64)

        self._bins = self.bins
        self._cnts = self.cnts

    def setBins(self, bins):

        self.bins = bins

        # If we don't set self._bins as well it will still refer to the previous
        # array and not the new one. In essence self._bins points to the memory
        # block and not the name self.bins

        self._bins = self.bins

    def getBins(self):
        return self.bins

    def setWeights(self, cnts):
        self.cnts = cnts
        self._cnts = self.cnts

    def getWeights(self):
        return self.cnts

    def setActiveBinCount(self, nActive):
        self.nActive = nActive

    def getActiveBinCount(self):
        return self.nActive

    def _assertCompiled(self):
        assert pyx.compiled

    @pyx.ccall
    @pyx.boundscheck(False)
    @pyx.wraparound(False)
    @pyx.initializedcheck(False)
    def _findLastLesserOrEqualIndex(self, point:pyx.double) -> pyx.int:

        idx:pyx.int = -1
        while True:
            if (self._bins[idx + 1] > point) or (idx + 1 == self.nActive):
                break
            else:
                idx += 1

        return idx

    @pyx.ccall
    @pyx.boundscheck(False)
    @pyx.wraparound(False)
    @pyx.initializedcheck(False)
    def _shiftRightAndInsert(self, idx:pyx.int, point:pyx.double, count:pyx.double) -> pyx.void:
        j: pyx.int

        for j in range(self.nActive - 1, idx, -1):
            self._bins[j+1] = self._bins[j]
            self._cnts[j+1] = self._cnts[j]

        self._bins[idx+1] = point
        self._cnts[idx+1] = count
        self.nActive += 1

    @pyx.cfunc
    @pyx.boundscheck(False)
    @pyx.wraparound(False)
    @pyx.initializedcheck(False)
    def _findMinimumDifference(self) -> pyx.int:
        k: pyx.int
        dB: pyx.double
        minK: pyx.int
        minDiff: pyx.double

        minK    = -1
        minDiff = 9e9
        # We don't want to merge the first or last bin because we want to maintain the
        # tails. It also solves the problem where we try and sample a point that is before
        # the first centroid.
        for k in range(1, self.nActive - 2):
            dB = self._bins[k+1] - self._bins[k]
            if dB < minDiff:
                minDiff = dB
                minK    = k

        return minK

    @pyx.ccall
    @pyx.boundscheck(False)
    @pyx.wraparound(False)
    @pyx.initializedcheck(False)
    def _shiftLeftAndOverride(self, idx: pyx.int) -> pyx.void:
        j: pyx.int
        for j in range(idx, self.nActive-1):
            self._bins[j] = self._bins[j+1]
            self._cnts[j] = self._cnts[j+1]

        self._bins[self.nActive-1] = 0
        self._cnts[self.nActive-1] = 0

        self.nActive -= 1

    @pyx.cfunc
    @pyx.boundscheck(False)
    @pyx.wraparound(False)
    @pyx.initializedcheck(False)
    def _sumWeights(self) -> pyx.double:
        som: pyx.double = 0
        i: pyx.int

        for i in range(self.nActive):
            som = som + self._cnts[i]

        return som

    def fit(self, x):
        for xx in x:
            self.add(xx)

    @pyx.ccall
    @pyx.boundscheck(False)
    @pyx.wraparound(False)
    @pyx.cdivision(True)
    @pyx.initializedcheck(False)
    def _add(self, point:pyx.double, count:pyx.double) -> pyx.void:

        idx:pyx.int = self._findLastLesserOrEqualIndex(point)

        if (idx >= 0) and self._bins[idx] == point:
            self._cnts[idx] += count
        else:
            self._shiftRightAndInsert(idx, point, count)

        if self.nActive > self.maxBins:
            k:pyx.int = self._findMinimumDifference()

            sumC:pyx.double = self._cnts[k+1] + self._cnts[k]

            self._bins[k] = (self._bins[k]*self._cnts[k] + self._bins[k+1]*self._cnts[k+1])
            self._bins[k] = self._bins[k] / sumC
            self._cnts[k] = sumC

            self._shiftLeftAndOverride(k+1)

    def centroids(self):
        return self.bins[:-1]

    def weights(self):
        return self.cnts[:-1]

    def add(self, point, count=1):
        self._add(point, count)

    def cdf(self, k):
        """

        Ted Dunning, Computing Extremely Accurate Quantiles Usings t-Digests
        """
        print('self.cdf(k)',k)
        som = 0
        c = self.bins
        m = self.cnts

        if k < self.lower():
            print('return 0.')
            return 0.
        elif k >= self.upper():
            print('return 1.')
            return 1.
        else:
            for i in range(self.nActive):
                if c[i] <= k < c[i+1]:
                    # We use the approach of Dunning here to improve interpolation when
                    # we have single weighted points.
                    if (m[i] > 1) & (m[i+1] > 1):
                        # Case I: Both points greater than one, normal interpolation.
                        yi   = som + m[i]/2
                        yi_n = yi + (m[i+1] + m[i]) / 2

                    elif (m[i] == 1) & (m[i+1] > 1):
                        # Case I: Both points greater than one, normal interpolation.
                        yi   = som
                        yi_n = yi + (m[i+1]) / 2

                    elif (m[i] > 1) & (m[i+1] == 1):
                        # Case I: Both points greater than one, normal interpolation.
                        yi   = som + m[i]/2
                        yi_n = yi + (m[i]) / 2
                    else:
                        yi   = som
                        yi_n = yi

                    g    = (yi_n - yi) / (c[i+1] - c[i])
                    yk   = g*(k - c[i]) + yi

                    return yk / self._sumWeights()

                else:
                    som += m[i]

            print(k)
            print(self.getBins())
            print(self.getWeights())


    @pyx.ccall
    @pyx.boundscheck(False)
    @pyx.cdivision(True)
    @pyx.wraparound(False)
    @pyx.initializedcheck(False)
    def pmf(self, kk: pyx.int) -> pyx.double:
        return self.cdf(kk + 0.5) - self.cdf(kk - 0.5)




    @pyx.ccall
    @pyx.boundscheck(False)
    @pyx.cdivision(True)
    @pyx.wraparound(False)
    @pyx.initializedcheck(False)
    def lower(self) -> pyx.int:
        return int(self._bins[0])

    @pyx.ccall
    @pyx.boundscheck(False)
    @pyx.cdivision(True)
    @pyx.wraparound(False)
    @pyx.initializedcheck(False)
    def upper(self) -> pyx.int:
        return int(self._bins[self.nActive - 1])

    def sample(self, size=1):
        p = np.random.rand()
        return int(round(self.quantile(p)))


    def quantile(self, p):
        # Using Regula Falsi
        a = self.lower()
        b = self.upper()

        fa = 0 - p
        fb = 1 - p

        iter = 0

        while ((b - a) > 0.1) and (iter < 100):
            c = (fb*a - fa*b) / (fb - fa)
            fc = self.cdf(c) - p

            if fc < 0:
                a  = c
                fa = fc
            elif fc > 0:
                b  = c
                fb = fc
            else:
                return c

            iter += 1

        return (b+a)/2

    def median(self):
        return self.quantile(0.5)

File: ranvar/cpp/digest.cpp
#include "digest.hpp"

#include <cmath>
#include <cstdlib>
#include <ctime>
#include <mutex>

namespace ranvar {

// ---------------------------------------------------------------------------
// Module-level random helpers
// ---------------------------------------------------------------------------

static std::once_flag rand_seed_flag_;

static void ensureSeeded() {
    std::call_once(rand_seed_flag_, []() {
        std::srand(static_cast<unsigned int>(std::time(nullptr)));
    });
}

double _rand() {
    ensureSeeded();
    return static_cast<double>(std::rand()) / static_cast<double>(RAND_MAX);
}

int _randint(double l, double h) {
    double l2  = l - 1.0;
    double out = std::ceil((h - l2) * _rand() + l2);
    if (out < l) out = l;
    return static_cast<int>(out);
}

double _randnorm(double mu, double stdev) {
    double x, y, s;
    do {
        x = 2.0 * _rand() - 1.0;
        y = 2.0 * _rand() - 1.0;
        s = x * x + y * y;
    } while (s >= 1.0);
    double z = x * std::sqrt(-2.0 * std::log(s) / s);
    return z * stdev + mu;
}

// ---------------------------------------------------------------------------
// Digest
// ---------------------------------------------------------------------------

Digest::Digest(int maxBins)
    : maxBins_(maxBins)
    , nActive_(0)
    , bins_(maxBins + 1, 0.0)
    , cnts_(maxBins + 1, 0.0)
{
}

// --- Private helpers -------------------------------------------------------

int Digest::findLastLesserOrEqualIndex(double point) const {
    int idx = -1;
    while (true) {
        if ((bins_[idx + 1] > point) || (idx + 1 == nActive_))
            break;
        ++idx;
    }
    return idx;
}

void Digest::shiftRightAndInsert(int idx, double point, double count) {
    for (int j = nActive_ - 1; j > idx; --j) {
        bins_[j + 1] = bins_[j];
        cnts_[j + 1] = cnts_[j];
    }
    bins_[idx + 1] = point;
    cnts_[idx + 1] = count;
    ++nActive_;
}

int Digest::findMinimumDifference() const {
    int    minK    = -1;
    double minDiff = 9e9;
    // Skip first and last centroids to preserve tails
    for (int k = 1; k < nActive_ - 2; ++k) {
        double dB = bins_[k + 1] - bins_[k];
        if (dB < minDiff) {
            minDiff = dB;
            minK    = k;
        }
    }
    return minK;
}

void Digest::shiftLeftAndOverride(int idx) {
    for (int j = idx; j < nActive_ - 1; ++j) {
        bins_[j] = bins_[j + 1];
        cnts_[j] = cnts_[j + 1];
    }
    bins_[nActive_ - 1] = 0.0;
    cnts_[nActive_ - 1] = 0.0;
    --nActive_;
}

double Digest::sumWeights() const {
    double sum = 0.0;
    for (int i = 0; i < nActive_; ++i)
        sum += cnts_[i];
    return sum;
}

void Digest::_add(double point, double count) {
    int idx = findLastLesserOrEqualIndex(point);

    if (idx >= 0 && bins_[idx] == point) {
        cnts_[idx] += count;
    } else {
        shiftRightAndInsert(idx, point, count);
    }

    if (nActive_ > maxBins_) {
        int    k    = findMinimumDifference();
        double sumC = cnts_[k] + cnts_[k + 1];
        bins_[k] = (bins_[k] * cnts_[k] + bins_[k + 1] * cnts_[k + 1]) / sumC;
        cnts_[k] = sumC;
        shiftLeftAndOverride(k + 1);
    }
}

// --- Public API ------------------------------------------------------------

void Digest::fit(const std::vector<double>& x) {
    for (double val : x)
        add(val);
}

void Digest::add(double point, double count) {
    _add(point, count);
}

std::vector<double> Digest::centroids() const {
    return std::vector<double>(bins_.begin(), bins_.begin() + nActive_);
}

std::vector<double> Digest::weights() const {
    return std::vector<double>(cnts_.begin(), cnts_.begin() + nActive_);
}

int Digest::lower() const {
    return static_cast<int>(bins_[0]);
}

int Digest::upper() const {
    return static_cast<int>(bins_[nActive_ - 1]);
}

double Digest::mean() const {
    double SxW = 0.0, W = 0.0;
    for (int i = 0; i < nActive_; ++i) {
        SxW += cnts_[i] * bins_[i];
        W   += cnts_[i];
    }
    return SxW / W;
}

double Digest::quantile(double p) const {
    if (p <= 0.0) return static_cast<double>(lower());
    if (p >= 1.0) return static_cast<double>(upper());

    double W  = sumWeights();
    double wi = 0.0;
    double w_ = p * W;

    for (int i = 0; i < nActive_ - 1; ++i) {
        double wGap;
        if (i == 0) {
            wGap = cnts_[i] + cnts_[i + 1] / 2.0;
        } else if (i == nActive_ - 1) {
            // unreachable: loop stops at nActive_-2, mirroring Python faithfully
            wGap = cnts_[i] / 2.0 + cnts_[i + 1];
        } else {
            wGap = cnts_[i] / 2.0 + cnts_[i + 1] / 2.0;
        }

        double wi_n = wi + wGap;

        if (wi <= w_ && w_ < wi_n) {
            double fraction = (w_ - wi) / wGap;
            return fraction * (bins_[i + 1] - bins_[i]) + bins_[i];
        }

        wi = wi_n;
    }

    return static_cast<double>(upper());
}

double Digest::cdf(double k) const {
    if (k <= static_cast<double>(lower())) return 0.0;
    if (k >= static_cast<double>(upper())) return 1.0;

    double som = 0.0;
    double W   = sumWeights();

    for (int i = 0; i < nActive_; ++i) {
        if (bins_[i] <= k && k < bins_[i + 1]) {
            double yi, yi_n;

            if (cnts_[i] > 1.0 && cnts_[i + 1] > 1.0) {
                yi   = som + cnts_[i] / 2.0;
                yi_n = yi  + (cnts_[i + 1] + cnts_[i]) / 2.0;
            } else if (cnts_[i] == 1.0 && cnts_[i + 1] > 1.0) {
                yi   = som;
                yi_n = yi + cnts_[i + 1] / 2.0;
            } else if (cnts_[i] > 1.0 && cnts_[i + 1] == 1.0) {
                yi   = som + cnts_[i] / 2.0;
                yi_n = yi  + cnts_[i] / 2.0;
            } else {
                yi   = som;
                yi_n = yi;
            }

            double g  = (yi_n - yi) / (bins_[i + 1] - bins_[i]);
            double yk = g * (k - bins_[i]) + yi;
            return yk / W;
        } else {
            som += cnts_[i];
        }
    }

    return 1.0;
}

double Digest::ccdf(double k) const {
    return 1.0 - cdf(k);
}

double Digest::dcdf(double k) const {
    if (k <= static_cast<double>(lower())) return 0.0;
    if (k >= static_cast<double>(upper())) return 0.0;

    double som = 0.0;
    double W   = sumWeights();

    for (int i = 0; i < nActive_; ++i) {
        if (bins_[i] <= k && k < bins_[i + 1]) {
            double yi, yi_n;

            if (cnts_[i] > 1.0 && cnts_[i + 1] > 1.0) {
                yi   = som + cnts_[i] / 2.0;
                yi_n = yi  + (cnts_[i + 1] + cnts_[i]) / 2.0;
            } else if (cnts_[i] == 1.0 && cnts_[i + 1] > 1.0) {
                yi   = som;
                yi_n = yi + cnts_[i + 1] / 2.0;
            } else if (cnts_[i] > 1.0 && cnts_[i + 1] == 1.0) {
                yi   = som + cnts_[i] / 2.0;
                yi_n = yi  + cnts_[i] / 2.0;
            } else {
                yi   = som;
                yi_n = yi;
            }

            double g = (yi_n - yi) / (bins_[i + 1] - bins_[i]);
            return g / W;
        } else {
            som += cnts_[i];
        }
    }

    return 0.0;
}

double Digest::dccdf(double k) const {
    return -dcdf(k);
}

Digest Digest::operator+(const Digest& other) const {
    double W1 = sumWeights();
    double W2 = other.sumWeights();
    int resultMaxBins = maxBins_ > other.maxBins_ ? maxBins_ : other.maxBins_;
    Digest result(resultMaxBins);
    if (W1 == 0.0 || W2 == 0.0) return result;
    for (int i = 0; i < nActive_; ++i) {
        double p1 = cnts_[i] / W1;
        for (int j = 0; j < other.nActive_; ++j) {
            double p2 = other.cnts_[j] / W2;
            result._add(bins_[i] + other.bins_[j], p1 * p2);
        }
    }
    return result;
}

double Digest::pmf(int kk) const {
    return cdf(kk + 0.5) - cdf(kk - 0.5);
}

int Digest::sample() {
    double p = _rand();
    return static_cast<int>(std::round(quantile(p)));
}

// --- Serialisation helpers -------------------------------------------------

int    Digest::getMaxBins()       const { return maxBins_; }
int    Digest::getActiveBinCount() const { return nActive_; }
void   Digest::setActiveBinCount(int n) { nActive_ = n; }
const  std::vector<double>& Digest::getBins() const { return bins_; }
const  std::vector<double>& Digest::getCnts() const { return cnts_; }
void   Digest::setBins(const std::vector<double>& bins) { bins_ = bins; }
void   Digest::setCnts(const std::vector<double>& cnts) { cnts_ = cnts; }

} // namespace ranvar
